trieste/models/interfaces.py

```diff
@@ -742,3 +743,159 @@ def covariance_with_top_fidelity(self, query_points: TensorType) -> TensorType:
         :return: The covariance with the top fidelity for the `query_points`, of shape [N, P]
         """
         raise NotImplementedError
+
+
+class EncodedProbabilisticModel(ProbabilisticModel):
+    """A probabilistic model with an associated query point encoder.
+
+    Classes that inherit from this (or the other associated mixins below) should implement the
+    relevant _encoded methods (e.g. predict_encoded instead of predict), to which the public
+    methods delegate after encoding their input. Take care to use the correct methods internally
+    to avoid encoding twice accidentally.
+    """
+
+    @property
+    @abstractmethod
+    def encoder(self) -> EncoderFunction | None:
```

Comment:
It seems a little strange to me that `encoder` and `encode` are public methods. But maybe this is necessary.

Reply:
Encoder is an abstract property to ensure that any encoded model has an actual encoder it can use.

Encode is a helper method for models that implement additional methods that need encoding (e.g. `DeepEnsemble.sample_ensemble`).